### PR TITLE
Nix CI: deploy smaller, toolbox-compatible Coq.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure.

Without switching yet the Nix setup to be entirely based on the Coq Nix Toolbox like #14435 does, this PR ensures that we are building and uploading to Cachix a version of Coq that will be directly useful for users, either for testing locally or for their CI.

Cf. the reworked documentation at https://github.com/coq/coq/wiki/Nix#coq-nix-toolbox.

There will therefore be two Nix jobs that run on `master` and release branches. One that builds the Coq nixpkgs package overridden to use the current version of the sources and that gets pushed to Cachix and one that builds a fuller package with tests and documentation, because this is still useful to detect regressions.

When https://github.com/NixOS/nixpkgs/pull/116690 is merged, it will be possible for users to do `nix-shell https://coq.inria.fr/nix/toolbox --arg override '{coq = "coq/coq-on-cachix:master";}'` instead of `nix-shell https://coq.inria.fr/nix/toolbox --arg override '{coq = "master";}'` to ensure that they always get the latest pre-built version.

cc @vbgl @CohenCyril

### CODE REVIEW

**Comment by CohenCyril** [comment]
@Zimmi48 I'm not sure I know how to see the resulting CI runs, could you help me there?

**Comment by Zimmi48** [comment]
Because this particular job was successful, it isn't explicitly listed in the reported status checks. But from the "GitLab CI pipeline" check, you can get a list with links of all the jobs that ran. As you can see `pkg:nix` did run successfully (log here: https://gitlab.com/coq/coq/-/jobs/1398088443). However, `pkg:nix:deploy` (the one that actually run a toolbox-based build) did not because it is set to only run on `master`. To properly test the part that only runs on `master`, I would have to set up my GitLab fork with the appropriate secret variable and push to `master` there.

**Comment by Zimmi48** [comment]
Let me know if you want me to do the additional testing.

**Comment by Zimmi48** [comment]
ping @CohenCyril

**Comment by CohenCyril** [comment]
> ping @CohenCyril

Sorry yes, I'm around but with high latency.


> To properly test the part that only runs on `master`, I would have to set up my GitLab fork with the appropriate secret variable and push to `master` there.

Could you do that? I'm unable to judge a CI script success just by reading it...

File: .gitlab-ci.yml
Comment (CohenCyril) [inline .gitlab-ci.yml:446]: I'm not sure I understand this line...

**Comment by CohenCyril** [comment]
> Could you do that? I'm unable to judge a CI script success just by reading it...

Unless it is ok to merge directly see failure (if happens) and fix later?

**Comment by Zimmi48** [comment]
> > To properly test the part that only runs on `master`, I would have to set up my GitLab fork with the appropriate secret variable and push to `master` there.
> 
> Could you do that? I'm unable to judge a CI script success just by reading it...

Yes. And in fact, most of the work (adding the secret variables) was actually already done for previous tests (a long time ago).

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml:446]: `CI_PROJECT_URL` and `CI_COMMIT_SHA` are set by the GitLab CI system. With them, we build the URL of the GitLab tarball archive for the commit being tested. And we build it with `nix-build`. Given that we haven't yet switched to using the Coq Nix Toolbox in this PR, this means that this will build the legacy `default.nix`, which includes the documentation and the test-suite.

**Comment by Zimmi48** [comment]
Testing before merging wasn't a bad idea since it revealed an issue (https://gitlab.com/Zimmi48/coq/-/jobs/1426758620):

```
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I), at /nix/store/dqi2xng1pngdasxfswa0ns6lq8mbf1cp-toolbox/.nix/fallback-config.nix:1:14
```

@CohenCyril Don't we want to change this line https://github.com/coq-community/coq-nix-toolbox/blob/7bc7046300baecfbb18b36571894032353ba5103/.nix/fallback-config.nix#L1 to not depend on the environment but on the nixpkgs version pinned in the toolbox?

**Comment by CohenCyril** [comment]
> Testing before merging wasn't a bad idea since it revealed an issue (https://gitlab.com/Zimmi48/coq/-/jobs/1426758620):
> 
> ```
> error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I), at /nix/store/dqi2xng1pngdasxfswa0ns6lq8mbf1cp-toolbox/.nix/fallback-config.nix:1:14
> ```
> 
> @CohenCyril Don't we want to change this line https://github.com/coq-community/coq-nix-toolbox/blob/7bc7046300baecfbb18b36571894032353ba5103/.nix/fallback-config.nix#L1 to not depend on the environment but on the nixpkgs version pinned in the toolbox?

Great !! Tmyes I was lazy, let's use the pinned version!

**Comment by Zimmi48** [comment]
Done in coq-community/coq-nix-toolbox#52. Testing this branch in https://gitlab.com/Zimmi48/coq/-/jobs/1432131594. It looks like this works (at least we are past the previous problem).

**Comment by Zimmi48** [comment]
OK, it worked. If this looks good to you @CohenCyril, I'll let you merge first coq-community/coq-nix-toolbox#52 and then this PR.

**Comment by Zimmi48** [comment]
@CohenCyril I'm waiting on you merging this PR to proceed with the next steps.

**Comment by Zimmi48** [comment]
ping @CohenCyril

File: .gitlab-ci.yml
Comment (CohenCyril) [reply .gitlab-ci.yml:449]: ok thanks for the explanation...

**Comment by CohenCyril** [comment]
@Zimmi48 I'm ok with merging this! (I guess that the errorhere: https://gitlab.com/Zimmi48/coq/-/jobs/1432131665 is spurious?)

**Comment by Zimmi48** [comment]
>I guess that the error here: https://gitlab.com/Zimmi48/coq/-/jobs/1432131665 is spurious?

Yes, that's because I cancelled the job that was building the Docker image and this job depends on it (but I didn't change anything to this deploy job, so there shouldn't be a need to test it).

>I'm ok with merging this!

OK, then you can go ahead and ask @coqbot to do it (cf. https://github.com/coq/coq/blob/master/CONTRIBUTING.md#merging-pull-requests).

**Review by CohenCyril — APPROVED** [review]
LGTM

**Comment by CohenCyril** [comment]
@coqbot: merge now

**Pull request merged**
